tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -11984,6 +11984,11 @@ SpirvEmitter::doUnaryExprOrTypeTraitExpr(const UnaryExprOrTypeTraitExpr *expr) {
     return nullptr;
   }
 
+  if (auto *constExpr = tryToEvaluateAsConst(expr)) {
+    constExpr->setRValue();
+    return constExpr;
+  }
+
   AlignmentSizeCalculator alignmentCalc(astContext, spirvOptions);
   uint32_t size = 0, stride = 0;
   std::tie(std::ignore, size) = alignmentCalc.getAlignmentAndSize(
```

Comment:
@jaebaek Thanks for fixing my issue!

Just one more question: Can we just call `AlignmentSizeCalculator` with a `scalar` memory layout instead of `void` here?

Reply:
yeah I think that's the more proper fix. Scalar memory layout allows non-scalar types to be aligned solely based on the size of their components (like C). Reference: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_EXT_scalar_block_layout.html#_description

Reply:
Good question. I think the `AlignmentSizeCalculator` with a `scalar` memory layout would also work.

In the example code,

```
struct StructA
{
    float3 one;
    float3 two;
};

uint foo()
{
    return sizeof(StructA);   // First case
}

[numthreads(64, 1, 1)]
void main()
{
    uint a = foo();
    uint b = sizeof(StructA);   // Second case
}
```

I just wanted to make the same result for `First case` and `Second case`.

[This code](https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L1395) is for the `Second case` and I wanted to reuse it.

In my opinion, even though both `tryToEvaluateAsConst()` and `AlignmentSizeCalculator` with a `scalar` memory layout will work, `tryToEvaluateAsConst()` would be a better option for this case.
We want to use what DXIL backend uses (as I mentioned `sizeof(some type)` does not rely on the Vulkan spec and we should mimic what the DXIL backend does) and `tryToEvaluateAsConst()` is what DXIL backend uses.

Reply:
Oh sorry. I misunderstood. Yeah, I agree that we have to change `void` to `scalar` memory layout in the line 11995 in addition to this change.